File: docker/worker-customcode-python/harness/honua_customcode_harness/harness.py
"""The custom-code harness entrypoint.

Flow (each step is a separately-testable function above):

  1. Load + validate job inputs (:mod:`jobspec`): repo_url, git_ref (40-hex SHA),
     entrypoint, deps_manifest, params_json, output_prefix, HONUA_BASE_URL,
     HONUA_JOB_TOKEN.
  2. Clone the user code at the pinned SHA and assert HEAD == SHA
     (:mod:`sourcefetch`).
  3. Restore user-declared extra deps (:mod:`deps`) — the SDK + geo stack are
     already baked into the image.
  4. Build the SCOPED HonuaClient from HONUA_JOB_TOKEN, then STRIP IMDS/token env
     (:mod:`sandbox`) BEFORE any user code is imported.
  5. Import the entrypoint (:mod:`loader`), build a :class:`GpContext`, call it.
  6. Upload returned artifacts to output_prefix (:mod:`upload`); map the
     :class:`GpResult` to a terminal exit code.

Exit codes: 0 = succeeded, 1 = tool returned/raised failure, 2 = harness/setup
error (bad inputs, clone/verify failure), 3 = job cancelled.
"""

from __future__ import annotations

import signal
import sys
import traceback
from collections.abc import Mapping
from pathlib import Path
from typing import Any

from .context import (
    CancellationToken,
    GpContext,
    GpResult,
    JobCancelled,
    OutputSink,
    OutputSizeExceeded,
    _LoggingProgressReporter,
    _StdLogger,
)
from .deps import DepsRestoreError, restore_requirements
from .jobspec import JobSpec, JobSpecError, load_job_spec
from .loader import EntrypointError, load_entrypoint
from .sandbox import (
    assert_credentials_stripped,
    build_scoped_client,
    strip_credential_env,
)
from .sourcefetch import SourceFetchError, clone_pinned
from .upload import ArtifactUploader

EXIT_OK = 0
EXIT_TOOL_FAILED = 1
EXIT_HARNESS_ERROR = 2
EXIT_CANCELLED = 3

DEFAULT_SOURCE_ROOT = Path("/work/src")
DEFAULT_WORKDIR = Path("/work/out")


def run(
    *,
    env: Mapping[str, str] | None = None,
    source_root: Path = DEFAULT_SOURCE_ROOT,
    workdir: Path = DEFAULT_WORKDIR,
    client_factory: Any = None,
    uploader_factory: Any = None,
    clone_fn=clone_pinned,
    restore_fn=restore_requirements,
    strip_env: bool = True,
) -> int:
    """Run one custom-code job end-to-end and return a process exit code.

    Most collaborators are injectable so the whole flow is testable offline.
    In production the defaults wire to git, pip, the real SDK, and boto3.
    """
    log = _StdLogger()
    cancellation = CancellationToken()
    _install_cancellation_handler(cancellation, log)

    # --- 1. Inputs ---------------------------------------------------------
    try:
        spec = load_job_spec(env)
    except JobSpecError as exc:
        log.warn(f"invalid job inputs: {exc}")
        return EXIT_HARNESS_ERROR

    # --- 2. Pinned source --------------------------------------------------
    try:
        log.info(f"cloning {spec.repo_url} @ {spec.git_ref}")
        clone_fn(spec.repo_url, spec.git_ref, source_root)
    except SourceFetchError as exc:
        log.warn(f"source fetch failed: {exc}")
        return EXIT_HARNESS_ERROR

    # --- 3. Restore extra deps (SDK + geo already baked in) ----------------
    try:
        manifest = restore_fn(source_root, spec.deps_manifest)
        if manifest is not None:
            log.info(f"restored extra deps from {manifest}")
    except DepsRestoreError as exc:
        log.warn(f"dependency restore failed: {exc}")
        return EXIT_HARNESS_ERROR

    # --- 4. Scoped client, THEN scrub credentials --------------------------
    try:
        client = build_scoped_client(
            spec.base_url, spec.job_token, client_factory=client_factory
        )
    except Exception as exc:  # noqa: BLE001 - surface SDK construction errors
        log.warn(f"failed to construct scoped client: {exc}")
        return EXIT_HARNESS_ERROR

    if strip_env:
        removed = strip_credential_env()
        log.info(f"stripped credential env before user code: {list(removed)}")
        # Invariant: nothing sensitive may remain visible to the tool.
        assert_credentials_stripped()

    # --- 5. Load entrypoint + run -----------------------------------------
    workdir.mkdir(parents=True, exist_ok=True)
    output = OutputSink(max_total_bytes=spec.output_max_bytes)
    context = GpContext(
        params=spec.params,
        inputs={},
        client=client,
        output=output,
        progress=_LoggingProgressReporter(log),
        log=log,
        cancellation=cancellation,
        workdir=workdir,
    )

    try:
        func = load_entrypoint(source_root, spec.entrypoint)
    except EntrypointError as exc:
        log.warn(f"entrypoint load failed: {exc}")
        return EXIT_HARNESS_ERROR

    try:
        result = func(context)
    except JobCancelled as exc:
        log.warn(f"job cancelled: {exc}")
        return EXIT_CANCELLED
    except Exception:  # noqa: BLE001 - any tool error is a tool failure
        log.warn("tool raised an unhandled exception:")
        log.warn(traceback.format_exc())
        return EXIT_TOOL_FAILED
    finally:
        _close_quietly(client)

    result = _coerce_result(result, log)
    if not result.ok:
        log.warn(f"tool reported failure: {result.message}")
        return EXIT_TOOL_FAILED

    # --- 6. Upload artifacts ----------------------------------------------
    try:
        uploader = _build_uploader(spec, uploader_factory)
        uploaded = uploader.upload(output.artifacts)
        for u in uploaded:
            log.info(f"uploaded {u.name} -> {u.uri} ({u.size_bytes} bytes)")
    except OutputSizeExceeded as exc:
        log.warn(f"output cap exceeded: {exc}")
        return EXIT_TOOL_FAILED
    except Exception as exc:  # noqa: BLE001 - upload failure is terminal
        log.warn(f"artifact upload failed: {exc}")
        return EXIT_HARNESS_ERROR

    log.info(f"job succeeded: {result.message or 'ok'}")
    return EXIT_OK


def _build_uploader(spec: JobSpec, uploader_factory: Any) -> Any:
    if uploader_factory is not None:
        return uploader_factory(spec.output_prefix)
    return ArtifactUploader(spec.output_prefix)


def _coerce_result(result: Any, log: Any) -> GpResult:
    if isinstance(result, GpResult):
        return result
    if result is None:
        # A tool that returns nothing but didn't raise is treated as success.
        return GpResult.succeeded()
    log.warn(
        f"tool returned {type(result)!r}, expected GpResult; treating as failure."
    )
    return GpResult.failed(f"entrypoint returned non-GpResult {type(result)!r}.")


def _install_cancellation_handler(token: CancellationToken, log: Any) -> None:
    def _handler(signum: int, _frame: Any) -> None:
        log.warn(f"received signal {signum}; requesting cancellation.")
        token.cancel()

    try:
        signal.signal(signal.SIGTERM, _handler)
        signal.signal(signal.SIGINT, _handler)
    except ValueError:  # pragma: no cover - not on main thread (tests)
        pass


def _close_quietly(client: Any) -> None:
    close = getattr(client, "close", None)
    if callable(close):
        try:
            close()
        except Exception:  # noqa: BLE001 - best-effort cleanup
            pass


def main(argv: list[str] | None = None) -> int:  # pragma: no cover - thin wrapper
    return run()


if __name__ == "__main__":  # pragma: no cover
    sys.exit(main())

File: docker/worker-customcode-dotnet/harness/src/Honua.CustomCode.Harness/Harness.cs
// Copyright (c) Honua. All rights reserved.
// Licensed under the Elastic License 2.0. See LICENSE in the project root.

using Honua.CustomCode.Sdk;

namespace Honua.CustomCode.Harness;

/// <summary>
/// The .NET custom-code harness orchestrator. Runs one job end-to-end and returns a
/// process exit code. This is the symmetric counterpart of the Python harness's
/// <c>run()</c> — same steps, same exit-code contract:
/// <list type="bullet">
/// <item>0 = succeeded</item>
/// <item>1 = tool returned/raised failure</item>
/// <item>2 = harness/setup error (bad inputs, clone/verify/build failure)</item>
/// <item>3 = job cancelled</item>
/// </list>
/// Most collaborators are injectable so the whole flow is testable offline.
/// </summary>
public sealed class Harness
{
    /// <summary>Exit code: the tool succeeded.</summary>
    public const int ExitOk = 0;

    /// <summary>Exit code: the tool returned or raised a failure.</summary>
    public const int ExitToolFailed = 1;

    /// <summary>Exit code: a harness/setup error (bad inputs, clone/verify/build failure).</summary>
    public const int ExitHarnessError = 2;

    /// <summary>Exit code: the job was cancelled.</summary>
    public const int ExitCancelled = 3;

    private readonly HarnessOptions _options;

    /// <summary>Creates a harness with the given (mostly injectable) options.</summary>
    /// <param name="options">The harness collaborators + directories.</param>
    public Harness(HarnessOptions? options = null)
    {
        _options = options ?? new HarnessOptions();
    }

    /// <summary>Run one custom-code job end-to-end.</summary>
    /// <param name="env">The job environment (defaults to the process environment).</param>
    /// <param name="cancellationToken">Cooperative cancellation for the whole run.</param>
    /// <returns>The terminal process exit code.</returns>
    public async Task<int> RunAsync(
        IReadOnlyDictionary<string, string?>? env = null,
        CancellationToken cancellationToken = default)
    {
        var log = _options.Logger ?? new HarnessLogger();

        // --- 1. Inputs ------------------------------------------------------
        JobSpec spec;
        try
        {
            spec = JobSpec.Load(env);
        }
        catch (JobSpecException ex)
        {
            log.Warn($"invalid job inputs: {ex.Message}");
            return ExitHarnessError;
        }

        // --- 2. Pinned source ----------------------------------------------
        try
        {
            log.Info($"cloning {spec.RepoUrl} @ {spec.GitRef}");
            _options.SourceFetch.ClonePinned(spec.RepoUrl, spec.GitRef, _options.SourceRoot);
        }
        catch (SourceFetchException ex)
        {
            log.Warn($"source fetch failed: {ex.Message}");
            return ExitHarnessError;
        }

        // --- 3. Build the user project (NTS + SDK already warm-cached) ------
        UserBuildResult build;
        try
        {
            log.Info($"building {spec.DepsManifest}");
            build = _options.ProjectBuilder.Build(_options.SourceRoot, spec.DepsManifest, _options.BuildOutput);
        }
        catch (UserBuildException ex)
        {
            log.Warn($"user project build failed: {ex.Message}");
            return ExitHarnessError;
        }

        // --- 4. Scoped client, THEN scrub credentials ----------------------
        object? client;
        try
        {
            client = CredentialSandbox.BuildScopedClient(spec.BaseUrl, spec.JobToken, _options.ClientFactory);
        }
        catch (Exception ex)
        {
            // Intentional catch-all: the injected client factory is arbitrary (tests
            // supply fakes, production wires a real HTTP client) and any construction
            // failure must degrade to a harness-error exit rather than crash the process.
            log.Warn($"failed to construct scoped client: {ex.Message}");
            return ExitHarnessError;
        }

        if (_options.StripEnvironment)
        {
            var removed = CredentialSandbox.StripCredentialEnv(_options.MutableEnv);
            log.Info($"stripped credential env before user code: {string.Join(", ", removed)}");
            // Invariant: nothing sensitive may remain visible to the tool.
            CredentialSandbox.AssertCredentialsStripped(_options.MutableEnv);
        }

        // --- 5. Load + activate the entrypoint, then run -------------------
        Directory.CreateDirectory(_options.WorkDir);
        var output = new OutputSink(spec.OutputMaxBytes);
        var context = new GpContext(
            spec.Params,
            new Dictionary<string, string>(),
            client,
            output,
            new LoggingProgressReporter(log),
            log,
            _options.WorkDir);

        IGeoprocessingTool tool;
        try
        {
            tool = _options.EntrypointLoader.Load(build.OutputDirectory, spec.Entrypoint);
        }
        catch (EntrypointException ex)
        {
            log.Warn($"entrypoint load failed: {ex.Message}");
            return ExitHarnessError;
        }

        // Scope the scoped-client disposal to the tool run itself: the credential-bearing
        // client must be released as soon as the tool finishes, not linger through the
        // artifact-upload step below.
        GpResult result;
        using (var disposableClient = client as IDisposable)
        {
            try
            {
                result = await tool.ExecuteAsync(context, cancellationToken).ConfigureAwait(false)
                         ?? GpResult.Succeeded();
            }
            catch (OperationCanceledException)
            {
                log.Warn("job cancelled");
                return ExitCancelled;
            }
            catch (Exception ex)
            {
                // Intentional catch-all: this boundary runs arbitrary user tool code, so any
                // exception type must be caught, logged in full, and mapped to the harness's
                // "tool failed" exit code rather than crash the harness process.
                log.Warn("tool raised an unhandled exception:");
                log.Warn(ex.ToString());
                return ExitToolFailed;
            }
        }

        if (!result.Ok)
        {
            log.Warn($"tool reported failure: {result.Message}");
            return ExitToolFailed;
        }

        // --- 6. Upload artifacts -------------------------------------------
        try
        {
            var uploader = _options.UploaderFactory is { } factory
                ? factory(spec.OutputPrefix)
                : new ArtifactUploader(spec.OutputPrefix);
            foreach (var uploaded in uploader.Upload(output.Artifacts))
            {
                log.Info($"uploaded {uploaded.Name} -> {uploaded.Uri} ({uploaded.SizeBytes} bytes)");
            }
        }
        catch (OutputSizeExceededException ex)
        {
            log.Warn($"output cap exceeded: {ex.Message}");
            return ExitToolFailed;
        }
        catch (Exception ex)
        {
            // Intentional catch-all: the uploader is injectable (S3 in production, a fake
            // in tests) and any upload failure must degrade to a harness-error exit rather
            // than crash the process after the tool has already succeeded.
            log.Warn($"artifact upload failed: {ex.Message}");
            return ExitHarnessError;
        }

        log.Info($"job succeeded: {result.Message ?? "ok"}");
        return ExitOk;
    }
}

/// <summary>
/// The injectable collaborators + directories for a <see cref="Harness"/> run. In
/// production the defaults wire to git, the .NET SDK, the real scoped Honua client, and
/// S3; tests inject fakes for every seam.
/// </summary>
public sealed class HarnessOptions
{
    /// <summary>The directory the user source is cloned into.</summary>
    public string SourceRoot { get; set; } = "/work/src";

    /// <summary>The directory the user project is built into.</summary>
    public string BuildOutput { get; set; } = "/work/build";

    /// <summary>The writable scratch directory handed to the tool.</summary>
    public string WorkDir { get; set; } = "/work/out";

    /// <summary>The git clone/verify collaborator.</summary>
    public SourceFetch SourceFetch { get; set; } = new();

    /// <summary>The user-project restore/build collaborator.</summary>
    public UserProjectBuilder ProjectBuilder { get; set; } = new();

    /// <summary>The entrypoint type loader/activator.</summary>
    public EntrypointLoader EntrypointLoader { get; set; } = new();

    /// <summary>An optional scoped-client factory override (tests inject a fake).</summary>
    public Func<string, string, object>? ClientFactory { get; set; }

    /// <summary>An optional uploader factory override (tests inject a fake).</summary>
    public Func<string, ArtifactUploader>? UploaderFactory { get; set; }

    /// <summary>An optional logger override.</summary>
    public IGpLogger? Logger { get; set; }

    /// <summary>Whether to strip the credential environment before user code (default true).</summary>
    public bool StripEnvironment { get; set; } = true;

    /// <summary>
    /// An optional in-memory environment the strip step operates on (tests). When null,
    /// the strip step mutates the real process environment.
    /// </summary>
    public IDictionary<string, string?>? MutableEnv { get; set; }
}
